contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,278 @@
+                    // Record current timestamp for each reward ineligible service
+                    curInfo.ts = block.timestamp;
+                }
+                // Record current service multisig nonce
+                curInfo.nonce = curNonce;
+
+                // Record the unstaked service Id index in the global set of staked service Ids
+                if (curServiceId == serviceId) {
+                    idx = i;
+                }
+            }
+
+            // Calculate each eligible service Id reward
+            uint256 tsLast = tsLastBalance;
+            // Calculate the maximum possible reward per service during the last deposit period
+            uint256 maxRewardsPerService = (rewardsPerSecond * (block.timestamp - tsLast)) / numServices;
+            // Traverse all the eligible services and calculate their rewards
+            for (uint256 i = 0; i < numServices; ++i) {
+                uint256 curServiceId = eligibleServiceIds[i];
+                ServiceInfo storage curInfo = mapServiceInfo[curServiceId];
+
+                // Calculate the reward up until now
+                // If the staking was longer than the deposited period, the service's timestamp is adjusted such that
+                // it is equal to at most the tsLastBalance of the last deposit happening during every _checkpoint() call
+                uint256 reward = rewardsPerSecond * (block.timestamp - curInfo.ts);
+                // Adjust the reward if it goes out of calculated max bounds
+                if (reward > maxRewardsPerService) {
+                    reward = maxRewardsPerService;
+                }
+
+                // Adjust the deposit balance
+                if (lastBalance >= reward) {
+                    lastBalance -= reward;
+                } else {
+                    // This situation must never happen
+                    // TODO: Fuzz this
+                    reward = lastBalance;
+                    lastBalance = 0;
+                }
+
+                // Add the calculated reward to the service info
+                curInfo.reward += reward;
+
+                // Adjust the starting ts for each service to a current timestamp
+                curInfo.ts = block.timestamp;
+            }
+            // Update the storage balance
+            balance = lastBalance;
+        }
+
+        // Record the current timestamp such that next calculations start from this point of time
+        tsLastBalance = block.timestamp;
+
+        emit Checkpoint(lastBalance);
+    }
+
+    /// @dev Unstakes the service.
+    /// @param serviceId Service Id.
+    function unstake(uint256 serviceId) external {
+        ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+        // Check for the service ownership
+        if (msg.sender != sInfo.owner) {
+            revert();
+        }
+
+        // Call the checkpoint and get the service index in the set of services
+        uint256 idx = _checkpoint(serviceId);
```

Comment:
We don't need this call. We simply need to update the two balances and then checkpoint is only called when people unstake. Keeps it simpler.

Also, I think we should have checkpoint being public so anyone can also just call it to update all reward allocations to services.

Also, I think there should be a getter method that virtualises the "would be" if one called checkpoint. Then the frontend can easily show current staking rewards for staked services.

Reply:
Agree with last two, but not the first one. First, if we want to maintain the set of staked services, we need to do the correct cleaning. Second, we can't just update balances without doing a checkpoint, as it would break the math of when different balances were added, see above.